tests/python/unittest/test_arith_solve_linear_inequality.py

```diff
@@ -0,0 +1,166 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+import random
+import sys
+import pytest
+import tvm
+from tvm import te, arith, ir, tir, testing
+
+
+def test_solve_system_of_inequalities():
+    seed = random.randrange(sys.maxsize)
+    print("\nThis test is intentionally non-deterministic, "
+          "if it fails please report it in github issue together with this seed {}\n".format(seed))
+    random.seed(seed)
+
+    def _check(variables, formulas, coef=(-5, 5), bounds=(-20, 20)):
+        vs = [te.var("x" + str(i)) for i in range(variables)]
+
+        fs = []
+        for i in range(formulas):
+            s1 = sum([v*random.randint(coef[0], coef[1]) for v in vs])
+            s1 += random.randint(coef[0], coef[1])
+            s2 = sum([v*random.randint(coef[0], coef[1]) for v in vs])
+            s2 += random.randint(coef[0], coef[1])
+            op = random.choice([tir.expr.EQ, tir.expr.LE, tir.expr.LT, tir.expr.GE, tir.expr.GT])
+            fs.append(op(s1, s2))
+
+        vranges = {v: tvm.ir.expr.Range(bounds[0], bounds[1] + 1) for v in vs}
+        before = te.all(tir.const(1, 'bool'), *fs)
+        after = arith._ffi_api.SolveInequalitiesAsCondition(vs, vranges, fs)
+        after = te.all(tir.const(1, 'bool'), *after)
+        testing.check_bool_expr_is_true(before == after, vranges)
```

Comment:
Thanks for the suggestion. It does find some bugs in the implementation, most related to round division, which might also exist in your branch, here's one example, 
```python
[(((x0 + (x1*2)) + 3) < ((x0 + (x1*-3)) + 1)),
((((x0*3) + (x1*-5)) + 1) >= ((x0*2) + -5)),
((((x0*-1) + (x1*-1)) + 2) <= ((x0 + (x1*4)) + 1))]
```
if we allow non-divisible results, it is solved to be
```
IntConstraintsTransform(
	IntConstraints([x0, x1], {x2: range(min=-20, ext=41), x1: range(min=-20, ext=41), x3: range(min=-20, ext=41), x0: range(min=-20, ext=41)}, [(((x0 + (x1*2)) + 3) < ((x0 + (x1*-3)) + 1)), ((((x0*3) + (x1*-5)) + 1) >= ((x0*2) + -5)), ((((x0*-1) + (x1*-1)) + 2) <= ((x0 + (x1*4)) + 1))])
	IntConstraints([x0.shifted, x1.shifted], {x0.shifted: range(min=0, ext=18), x1.shifted: range(min=0, ext=7)}, [(36 <= (((x1.shifted*5) + (floordiv((37 - (x1.shifted*5)), 2)*2)) + (x0.shifted*2))), ((floordiv((37 - (x1.shifted*5)), 2) + x0.shifted) <= 20)])
	{x0: (x0.shifted + floordiv((37 + (x1.shifted*-5)), 2)), x1: (x1.shifted + -7)}
	{x0.shifted: ((x0 - floordiv((0 - (x1*5)), 2)) - 1), x1.shifted: (x1 + 7)}
)
```
I modified the `FindBestRange` to skip non-divisible cases,
```cpp
if (analyzer.CanProve(floormod(low, coef) == 0)) {
      // we need to be very careful with rounding
      // as it could be wrong when we have equations.
      // equations can come from
      // 1. when it is a single point, i.e., extent == 1.
      // 2. when var is substituted by another var in deskew range.
      best_lower = low_divided;
      best_diff_over = diff_over;
}
```

Reply:
Nice! I reproduced the bug on my old branch and debugged it a little bit. I think the real cause is a bug in the canonical simplifier, because on my branch it rewrites
`((x0 - floordiv((x1*-5), 2)) + -1)` into `((x0 - ((x1*-5)/2)) - 1)` given ranges
```{x0: range(min=-20, ext=41), x1: range(min=-20, ext=41)}```. I'm not familiar with the internals of the canonical simplifier, so I don't know how to fix it properly (seems like when an expr is converted to SplitExpr, its div_mode is always initialized as kTruncDiv, which is wrong). Could you take a look?

Reply:
It does not simplified to truncdiv in my branch though. I think the problem is rather like, `2x > 1` can be rounded to be `x > 1`, yet `2x == 1` cannot be rounded as `x == 1` or `x == 0`. The equations need to be specially handled.

Reply:
`2x == 1` should be transformed to `x >= ceildiv(1, 2) = 1` and `x <= floordiv(1, 2) = 0` which gives the empty set, so I don't think that equations should be problematic.

I checked out your branch and the bug I mentioned does seem to be fixed indeed. However your example results in the following failure (when the workaround for equations is off):
```
AssertionError: Expression ((((((x0: int32 + 1) - (x1: int32*3)) <= (((x1*2) + x0) + 3)) || ((((x0*3) + 1) - (x1*5)) < ((x0*2) - 5))) || ((((x1*4) + x0) + 1) < ((2 - x0) - x1))) || ((((1 <= ((x1*5) + (x0*2))) && (x0 <= 20)) && ((1 <= (x0 - ((x1*-5) / 2))) && ((x0 - ((x1*-5) / 2)) < 19))) && ((-7 <= x1) && (x1 < 0))))                                                                      
is not true on {x1: range(min=-20, ext=41), x0: range(min=-20, ext=41)}
Counterexample: x0: int32 = 18, x1: int32 = -7
```
But the expression is true for the given values if I'm not mistaken. Could be a bug in the testing function or a miscompile.

(BTW, I see that trunc division appeared during simplification which probably shouldn't happen. Although it might not be the cause of the bug, it still seems bad because truncated division is harder to deal with and may cause poor performance of some transformations)

Reply:
I think you're right. It is something wrong with the testing function and compilation flow.
during the testing & compilation, simplification is applied and seems to mistakenly transform floordiv to truncdiv, here is one example,

```python
>>> import tvm
>>> from tvm import te
>>> from tvm.tir import floordiv
>>> ana = tvm.arith.Analyzer()
>>> x0, x1 = te.var("x0"), te.var("x1")
>>> expr = (((x0 - floordiv((0 - (x1*5)), 2)) - 1) + floordiv((37 + ((x1 + 7)*-5)), 2))
>>> expr_sim = ana.simplify(expr)
>>> expr_sim
((floordiv((0 - (x1*5)), 2) + x0) - ((x1*-5)/2))
>>> ana.simplify(tvm.tir.stmt_functor.substitute(expr, {x0: 1, x1: 3}))
1
>>> ana.simplify(tvm.tir.stmt_functor.substitute(expr_sim, {x0: 1, x1: 3}))
0  # <- showing inconsistent results between expr and expr_sim
```

I can confirm this is the part of logic in canonical simplification https://github.com/apache/incubator-tvm/blob/master/src/arith/canonical_simplify.cc#L757

@tqchen do you have any idea about it?

Reply:
https://github.com/apache/incubator-tvm/pull/5922

Reply:
The issue resolves after merging the canonical simplify fix. I also remove the workaround for equations.